### PR TITLE
feat(eslint-plugin): [prefer-string-starts-ends-with] add allowSingleElementEquality option

### LINKED ISSUE #7140: Enhancement: [prefer-string-starts-ends-with] Option to ignore [0] === cases?
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3gNyhAriAFzyYwCWKA5gNwBQd5AZvABS4EgDaADALrwAvMJx5CvPgDoMSAKoAHeSBgBhKGhCsAlFvgBvOvHgB6Y-EkW6AXwbM2HQpMywMaAOrkMAC3Zju-aTlFZTUNbR19QxMzC0lrOiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WluXxUmAc374AhtHzJ6TACYyA7pXwALdFF7QA9tEjgwAXxCGgA

### Description

Coming over from https://github.com/typescript-eslint/typescript-eslint/pull/7138#discussion_r1241171363: when checking whether a string's `[0]` member equals a modification of `[0]`, the rule's preference ends up being even more code:

```ts
declare const value: string;

if (value[0] === value[0].toUpperCase()) { /* ... */ } // before

if (value.startsWith(value[0].toUpperCase())) { /* ... */ } // after
```

Agreeing with @bradzacher:

> I actually think that the previous form is clearer for this specific case.
> [It's also substantially faster (not that it matters at the this speed... but it is twice as fast - ~1.2B ops/s vs ~2.4B ops/s)](https://jsbench.me/buljbfvcn4/1). 
>
> I wonder if we should add an option to skip this case from the rule?

It's not yet clear to me how we would categorize and describe "this case". Maybe it's that we don't want to flag cases that are just checking whether `value[0] ===` something? In that case, maybe the option would be named something like `checkFirstElementEquality`?

### Fail

```typescript
n/a
```

### Pass

```typescript
n/a
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7140
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Adds a new `allowSingleElementEquality` option to the rule that's allowed to be `'never'` (default) or `'always'`.

Doesn't yet add a third option for case matching as described in the issue. Is that ok to leave as a followup?

Also removes the auto-generation of tests from `prefer-string-starts-ends-with.test.ts` per #6887. The test file predates that preference.

### CODE REVIEW

**Comment by JoshuaKGoldberg** [comment]
On mobile, let's assume I'll update the snapshots before merge :)

File: packages/eslint-plugin/docs/rules/prefer-string-starts-ends-with.md [hunk 1]
```diff
@@ -52,6 +52,26 @@ foo.startsWith('bar');
...
+## Options
+
+### `allowSingleElementEquality`
+
+If switched to `'always'`, the rule will allow equality checks against the first or last character in a string.
+This can be preferable in projects that don't deal with special character encodings and prefer a more succinct style.
+
+The following code is considered incorrect by default, but is allowed with `allowSingleElementEquality: 'always'`:
+
+```ts
```
Comment (auvred) [inline packages/eslint-plugin/docs/rules/prefer-string-starts-ends-with.md:66]:
```suggestion
```ts option='{ "allowSingleElementEquality": "always" }' showPlaygroundButton
```

What do you think about adding "Open in Playground" button here? :slightly_smiling_face:

**Review by auvred — APPROVED** [review]
Looks great! :fire:

File: packages/eslint-plugin/docs/rules/prefer-string-starts-ends-with.md (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/prefer-string-starts-ends-with.md:66]: Oh good point. I wonder if we should add in something to make sure we always remember to ask for that (I don't)... or make it more automatic. 🤔

**Review by bradzacher — APPROVED** [review]

**Pull request merged**